Fix #1342: Add frontend for editing a challenge phase details.

isht3 opened the pull request:
Fixes: #1342 


1.

![](https://image.prntscr.com/image/2uqDcBFVQ-_vdUAlxPWRXg.png)

2.

![](https://image.prntscr.com/image/04C3vYEPQt2rAntN9RzQAA.png)

3.

![](https://image.prntscr.com/image/I9Rmu3hyQ2GTQHXofd_UqQ.png)

4.

![](https://image.prntscr.com/image/uq79PJalSdm0Te4By9SSjw.png)

5.

![](https://image.prntscr.com/image/RAK0WFMSTmyuf33NUJJIRw.png)

6.

![](https://image.prntscr.com/image/_M878oAeQdW73HiKKazFBg.png)

7.

![](https://image.prntscr.com/image/YmWVCwGpSHGaV7ppjtQ2Gw.png)

8.

![](https://image.prntscr.com/image/7OIF9RDUQlmb8ic0HuAhpA.png)

isht3 commented:
@RishabhJain2018, @aka-jain Do let me know what all changes to make in the UI! :smile:

deshraj commented:
@isht3: some comments:

1. Can you please make the dialog box wider? It's hard for anyone to see and edit the challenge phase description as of now. 
2. `Maximum number of Submissions/Day` cannot be negative.
3. `Maximum total submissions` cannot be negative. 
4. Replace "Please enter the challenge phase name to update the test annotation file." with "Please enter the challenge phase name i.e **<CHALLENGE_PHASE_NAME_HERE>** to update the test annotation file."

Otherwise looks good.

isht3 commented:
@deshraj I think the regex `ng-pattern="/^\+?(0|[1-9]\d*)$/"` used in both submission inputs only allow positive integers, I'll give it a quick testing and I'll let you know.

isht3 commented:
@deshraj I did the changes, should I update the screenshots?

RishabhJain2018 commented:
@isht3 yes :-)

isht3 commented:
@deshraj I've made the requested changes! :smile:

isht3 commented:
@aka-jain Can you please review this? 😄

aka-jain commented on frontend/src/views/web/challenge/phases.html:
```diff
@@ -1,19 +1,27 @@
-            <div class="col s12">
-                <span class="fs-18 text-med-black"><strong>{{item.name}}</strong></span>
+            <div class="col s11 m11">
+                <div class="col s12">
```
Wrap col in row please. This is bad practice for grid implementation.
FYI: Cover col-12 in A ROW. For every new row wrap col in 'row' class. For example
```
<div class="row">
    <div class="col s8">
         <div class="row">
               <div class="col s12"></div>
         </div>
     </div>
    <div class="col sm4"></div>
</div>
```
Hope you get it.

aka-jain commented on frontend/src/js/controllers/challengeCtrl.js:
```diff
@@ -1415,6 +1415,82 @@
+            vm.sanity_check_pass = false;
+        };
+
+        vm.sanityCheckPass = function() {
```
Create a service for 'SanityCheck' please

isht3 replied:
Hi @aka-jain! I didn't quite understand how I could convert this function into a service because it uses variables in the within the `challengeCtrl`, but I did manage to get rid of the other function and reduce the total number of lines of code.

aka-jain commented:
Commented, please check ! @isht3

isht3 commented:
@aka-jain Thanks a tonne! I'll work on getting this fixed.

isht3 commented:
@aka-jain One problem I'm facing is while uploading the `test_annotation` file to the request data, when I send it in the raw format (What I'm doing now) I get this response `{"test_annotation":["The submitted data was not a file. Check the encoding type on the form."]}`, am I supposed to use `FormData` for sending files or should changes be added to the backend to support this?

isht3 commented:
@aka-jain Any idea? :smile:

aka-jain commented:
@RishabhJain2018 please take a look here.

deshraj commented:
@isht3 it seems like you are not using the 'Content-Type: application/x-www-form-urlencoded' header with the request that you are making. Also, submit the data as `form-data` and it will work properly.

isht3 commented:
@deshraj I did the required changes, I managed to get rid of that error, now it shows `"detail":"Unsupported media type \"text/plain;charset=UTF-8\" in request."}`. I'm using `test_annotation.txt` to test, should I be using JSON?

deshraj commented:
Yeah. Please use JSON.

RishabhJain2018 commented on frontend/src/js/controllers/challengeCtrl.js:
```diff
@@ -1415,6 +1415,88 @@
 
+        vm.challengePhaseDialog = function(ev, phase) {
+            vm.page.challenge_phase = phase;
+            vm.phase_start_date = new Date(phase.start_date);
```
Can we please use `CamelCase` in the variable names.

isht3 replied:
Done!

isht3 replied:
@RishabhJain2018 Done!

RishabhJain2018 commented on frontend/src/js/controllers/challengeCtrl.js:
```diff
@@ -1415,6 +1415,88 @@
+                    parameters.data.test_annotation = formData;
+                }
+                parameters.token = userKey;
+                parameters.headers = {'Content-Type': 'application/x-www-form-urlencoded'};
```
I think there is no need to add the header specifically. Please check `services.js`, we have a service for it.

isht3 replied:
@RishabhJain2018 Is it `sendRequest`? How do I use it?

RishabhJain2018 commented:
@isht3 Is the file uploading problem resolved? If not, then I have a solution for it. Please update me on this.

isht3 commented:
@RishabhJain2018 I couldn't resolve it. Could you tell me the solution? I'll incorporate it, in this PR. 😄

isht3 commented:
@RishabhJain2018 Can I split this PR into two? Since the Challenge Phase editing is done (Except `test_annotation`) and only the file upload is left can I split the File Upload into another PR, so it will be more clear to work with?

RishabhJain2018 commented on frontend/src/js/controllers/challengeCtrl.js:
```diff
@@ -1415,6 +1415,88 @@
+                    parameters.data.test_annotation = formData;
+                }
+                parameters.token = userKey;
+                parameters.headers = {'Content-Type': 'application/x-www-form-urlencoded'};
```
Please remove the headers.

RishabhJain2018 commented on frontend/src/js/controllers/challengeCtrl.js:
```diff
@@ -1415,6 +1415,88 @@
+                    }
+                };
+                utilities.showLoader();
+                utilities.sendRequest(parameters);
```
Please change this to `utilities.sendRequest(parameters, 'header', 'upload');`

RishabhJain2018 commented on frontend/src/js/controllers/challengeCtrl.js:
```diff
@@ -1415,6 +1415,88 @@
+                vm.challengePhaseId = vm.page.challenge_phase.id;
+                var parameters = {};
+                parameters.url = "challenges/challenge/" + vm.challengeId + "/challenge_phase/" + vm.challengePhaseId;
+                parameters.method = 'PATCH';
```
Now, create a `FormData` object here and pass the values to it one by one, like
`var formData = new FormData();
formData.append("name", vm.page.challenge_phase.name);
formData.append("description", vm.page.challenge_phase.description);` and so on.

Finally after taking all the fields, please assign this to parameters using
`parameters.data = formData`

Please try this and I think it will work. Let me know if ypou still have some doubts or queries.

RishabhJain2018 commented on frontend/src/js/controllers/challengeCtrl.js:
```diff
@@ -1415,6 +1415,88 @@
+                var parameters = {};
+                parameters.url = "challenges/challenge/" + vm.challengeId + "/challenge_phase/" + vm.challengePhaseId;
+                parameters.method = 'PATCH';
+                parameters.data = vm.page.challenge_phase;
```
Please remove this line.

RishabhJain2018 commented on frontend/src/js/controllers/challengeCtrl.js:
```diff
@@ -1415,6 +1415,88 @@
+                parameters.method = 'PATCH';
+                parameters.data = vm.page.challenge_phase;
+                if (vm.testAnnotationFile){
+                    var formData = new FormData();
```
Create the object above in the starting. Don't create a new object here, use the previous one!

isht3 commented:
@RishabhJain2018 Thank you SOOOOOOOO much! It worked finally!!! I'll update the PR now. :smile:
